contributors/design-proposals/changing-merge-key.md

```diff
@@ -0,0 +1,670 @@
+# Support Sending Merge Key in the Patch by Clients
+
+## Abstract
+
+Include the merge key in a patch directive sending by the clients.
+
+## Background
+
+Strategic Merge Patch is covered in this [doc](https://github.com/kubernetes/community/blob/master/contributors/devel/strategic-merge-patch.md).
+In Strategic Merge Patch, Merge Key is the key to identify the entries in the list of non-primitive types.
+It must always be present and unique to perform the merge on the list of non-primitive types,
+and will be preserved.
+
+The merge key exists in the struct tag (e.g. in [types.go](https://github.com/kubernetes/kubernetes/blob/5a9759b0b41d5e9bbd90d5a8f3a4e0a6c0b23b47/pkg/api/v1/types.go#L2831))
+and the [OpenAPI spec](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json).
+
+`patchMergeKey` tags always exist with `patchStrategy: "merge"`.
+`patchStrategy: "merge"` may exist without `patchMergeKey` tags for list of primitives.
+
+We are supporting multi-fields merge key in [#610](https://github.com/kubernetes/community/pull/610).
+This proposal depends on #610.
+
+## Motivation
+
+There are some existing APIs that are using merge keys incorrectly.
+
+The current existing APIs requires a single field that uniquely identifies each element in a list.
+For some element Kinds, the identity is defined using multiple fields.
+An [example](https://github.com/kubernetes/kubernetes/issues/39188) is the service.spec.ports,
+which is identified by both `protocol` and `port`.
+Therefore, we should change the merge key to multi-fields.
+
+There are some existing APIs that put the merge key tag at the wrong place.
+We should also correct them.
+
+## Goal
+
+Propose a solution of changing the merge keys for existing APIs without breaking backward compatibility.
```

Comment:
We need to take a step back and think about this systematically.

Reply:
I think we should consider building a more principled form of the patch syntax. We keep making ad-hoc fixes.

Reply:
Yeah, the existing RFCs have their limitations but evolving one of those with the improvements from SMR seems like a stronger path than retrofitting into existing SMR.